### PR TITLE
Fixed #24128 -- Made admindocs TemplateDetailView respect template_loaders.

### PR DESCRIPTION
**Pull request by e11bits**
This is the proposed PR #7434 by Kris Avi. I rebased it onto the current main branch and created a ~draft~ PR for it.

I had to remove some duplicate code and revert some outdated changes. I left all the commits in, but could squash them if needed for review.

@ngnpope and @timgraham are the reviewers of the former PR and might want to have a look at this again as well?

This is my third PR for Django, so please advise as you see fit.

### CODE REVIEW

File: django/template/loaders/cached.py [hunk 1]
```diff
@@ -18,9 +18,7 @@ def __init__(self, engine, loaders):
         super().__init__(engine)
 
     def get_dirs(self):
-        for loader in self.loaders:
-            if hasattr(loader, "get_dirs"):
-                yield from loader.get_dirs()
+        return [dir for loader in self.loaders for dir in loader.get_dirs() if loader]
```
Comment (kezabelle) [inline django/template/loaders/cached.py:21]: To the best of my _recollection_, `get_dirs` isn't **required** on a loader, hence the existing code doing the `hasattr` check, which was added in 29845ecf6999e9e31b56a82229d885f8dfb99e04 as part of [ticket #25791](https://code.djangoproject.com/ticket/25791) which is _more recent_ than the PR you've rebased -- as such, I'd assume this bit can/should be reverted/unnecessary?

File: django/contrib/admindocs/views.py [hunk 2]
```diff
@@ -404,8 +404,15 @@ def get_context_data(self, **kwargs):
             # Non-trivial TEMPLATES settings aren't supported (#24125).
             pass
         else:
-            # This doesn't account for template loaders (#24128).
-            for index, directory in enumerate(default_engine.dirs):
+            directories = list(default_engine.dirs)
+            for loader in default_engine.template_loaders:
+                directories.extend(loader.get_dirs())
```
Comment (kezabelle) [inline django/contrib/admindocs/views.py:409]: _Assuming_ (🤷) my other remark about `get_dirs` is correct, I'd presume this line needs a `if hasattr(loader, "get_dirs")` preceding it...

File: django/template/loaders/cached.py (hunk 1 above)
Comment (e11bits) [reply django/template/loaders/cached.py:21]: Ah, yes. I was making the change based on 57864dc125a8ae1c624da5cbb6b1cc3d1171c4af

File: django/contrib/admindocs/views.py (hunk 2 above)
Comment (e11bits) [reply django/contrib/admindocs/views.py:409]: Yes, I think you are right. 👍🏼

Thanks, that you took the time to have a look at this PR! 🙏🏼

**Comment by e11bits** [comment]
I added one more commit " Removes reading template contents not used in template_detail.html" . Strictly this has nothing to do with the bug ticket. I can remove it again if necessary.

**Comment by e11bits** [comment]
Squashed all commits but the last one

**Comment by felixxm** [comment]
> I added one more commit " Removes reading template contents not used in template_detail.html" . Strictly this has nothing to do with the bug ticket. I can remove it again if necessary.

Yes. this is unrelated and backward incompatible change. Please revert.

File: django/contrib/admindocs/views.py [hunk 3]
```diff
@@ -404,21 +404,25 @@ def get_context_data(self, **kwargs):
             # Non-trivial TEMPLATES settings aren't supported (#24125).
             pass
         else:
-            # This doesn't account for template loaders (#24128).
-            for index, directory in enumerate(default_engine.dirs):
+            directories = list(default_engine.dirs)
+            for loader in default_engine.template_loaders:
+                if hasattr(loader, "get_dirs"):
+                    directories.extend(loader.get_dirs())
```
Comment (felixxm) [inline django/contrib/admindocs/views.py:410]: Is it worth going back and forward on the list of directories? We need to evaluate the entire list, so we can do this at the beginning and avoid doing this twice, for example:
```diff
diff --git a/django/contrib/admindocs/views.py b/django/contrib/admindocs/views.py
index bd9cd952df..f4c5509390 100644
--- a/django/contrib/admindocs/views.py
+++ b/django/contrib/admindocs/views.py
@@ -407,13 +407,11 @@ class TemplateDetailView(BaseAdminDocsView):
             directories = list(default_engine.dirs)
             for loader in default_engine.template_loaders:
                 if hasattr(loader, "get_dirs"):
-                    directories.extend(loader.get_dirs())
-            # Go backwards over directories and ignore duplicates on the way. Template
-            # uses dictsort to show directories in right order and with right index.
-            index = 0
-            while directories:
-                if not (directory := directories.pop()) or directory in directories:
-                    continue
+                    for dir_ in loader.get_dirs():
+                        if dir_ not in directories:
+                            directories.append(dir_)
+
+            for index, directory in enumerate(directories):
                 template_file = Path(safe_join(directory, template))
                 templates.append(
                     {
@@ -422,7 +420,6 @@ class TemplateDetailView(BaseAdminDocsView):
                         "order": index,
                     }
                 )
-                index -= 1
         return super().get_context_data(
             **{
                 **kwargs,

```

**Review by felixxm — COMMENTED** [review]
@e11bits Thanks :+1:

File: django/contrib/admindocs/views.py (hunk 3 above)
Comment (e11bits) [reply django/contrib/admindocs/views.py:410]: True, that's better.

**Comment by e11bits** [comment]
I applied the recommendations

**Review by felixxm — COMMENTED** [review]
@e11bits Thanks for polishing this patch :+1: 

@krisavi Welcome aboard :boat:

**Pull request merged**